src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -1001,6 +974,139 @@ public boolean connected() {
+                throw new MqttException(String.format("%s must be greater than %s",
+                        MQTT_KEEP_ALIVE_TIMEOUT_KEY, MQTT_PING_TIMEOUT_KEY));
+            }
+            mqttTimeoutMs = Coerce.toInt(
+                    mqttConfig.getOrDefault(MQTT_OPERATION_TIMEOUT_KEY, DEFAULT_MQTT_OPERATION_TIMEOUT));
+            socketTimeoutMs = Coerce.toInt(
+                    mqttConfig.getOrDefault(MQTT_SOCKET_TIMEOUT_KEY, DEFAULT_MQTT_SOCKET_TIMEOUT));
+            port = Coerce.toInt(mqttConfig.getOrDefault(MQTT_PORT_KEY, DEFAULT_MQTT_PORT));
+
+            if (nucleusConfig.containsKey(DEVICE_PARAM_IOT_DATA_ENDPOINT)) {
+                endpoint = Coerce.toString(nucleusConfig.get(DEVICE_PARAM_IOT_DATA_ENDPOINT));
+            } else {
+                endpoint = Coerce.toString(deviceConfiguration.getIotDataEndpoint());
+            }
+
+            synchronized (httpProxyLock) {
+                Map<String, Object> networkProxy = deviceConfiguration.getNetworkProxyNamespace().toPOJO();
+                if (nucleusConfig.containsKey(DEVICE_NETWORK_PROXY_NAMESPACE)) {
+                    networkProxy = (Map<String, Object>) nucleusConfig.get(DEVICE_NETWORK_PROXY_NAMESPACE);
+                }
+                Map<String, Object> proxy = (Map<String, Object>) networkProxy.get(DEVICE_PROXY_NAMESPACE);
+
+                String proxyUrl = Coerce.toString(proxy.get(DEVICE_PARAM_PROXY_URL));
+                if (Utils.isEmpty(proxyUrl)) {
+                    return;
+                }
+                String proxyUsername = ProxyUtils.getProxyUsername(proxyUrl,
+                        Coerce.toString(proxy.getOrDefault(DEVICE_PARAM_PROXY_USERNAME, "")));
+                String proxyPassword = ProxyUtils.getProxyPassword(proxyUrl,
+                        Coerce.toString(proxy.getOrDefault(DEVICE_PARAM_PROXY_PASSWORD, "")));
+
+                httpProxyOptions = ProxyUtils
+                        .getHttpProxyOptions(proxyUrl, proxyUsername, proxyPassword, proxyTlsContext);
+                if (httpProxyOptions != null) {
+                    String noProxy = Coerce
+                            .toString(networkProxy.getOrDefault(DEVICE_PARAM_NO_PROXY_ADDRESSES, ""));
+                    useProxy = true;
+                    // Only use the proxy when the endpoint we're connecting to is not in the NoProxyAddress list
+                    if (Utils.isNotEmpty(noProxy) && Utils.isNotEmpty(endpoint)) {
+                        useProxy = Arrays.stream(noProxy.split(",")).noneMatch(endpoint::matches);
+                    }
+                }
+            }
+        }
+
+        private AwsIotMqttConnectionBuilder createAwsIotMqttConnectionBuilder(ClientBootstrap clientBootstrap)
+                throws MqttException {
+            AwsIotMqttConnectionBuilder builder;
+            try {
+                builder = securityService.getDeviceIdentityMqttConnectionBuilder();
+            } catch (MqttConnectionProviderException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While re-throwing the caught exception with modifications , information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)

Suggested remediation:
Consider passing the exception object either as  &#34;rootCause&#34; or as an inner exception parameter, to the constructor of the new exception before throwing new exception. This will help keep stack trace.
```
@@ -1084,3 +1084,3 @@
             } catch (MqttConnectionProviderException e) {
-                throw new MqttException(e.getMessage());
+                throw new MqttException(e.getMessage(), e);
             }

```